### ISSUE TITLE
Cannot import Mobileclient in python app engine

### ISSUE THREAD

**Issue by clocksmith** [simon-weber/gmusicapi#324]
I first encountered a protobuf issue, but this fixed it:
https://www.simonmweber.com/2013/06/18/python-protobuf-on-app-engine.html

Now it cannot find pwd. I read that it might have something to do with requests, and I have tried multiple version with no luck.

```
File "/Users/*/google-cloud-sdk/platform/google_appengine/google/appengine/runtime/wsgi.py", line 240, in Handle
    handler = _config_handle.add_wsgi_middleware(self._LoadHandler())
  File "/Users/*/google-cloud-sdk/platform/google_appengine/google/appengine/runtime/wsgi.py", line 299, in _LoadHandler
    handler, path, err = LoadObject(self._handler)
  File "/Users/*/google-cloud-sdk/platform/google_appengine/google/appengine/runtime/wsgi.py", line 85, in LoadObject
    obj = __import__(path[0])
  File "/Users/*/git/*/main.py", line 30, in <module>
    from gmusicapi import Mobileclient
  File "libs/gmusicapi/__init__.py", line 9, in <module>
    from gmusicapi.clients import Webclient, Musicmanager, Mobileclient
  File "libs/gmusicapi/clients/__init__.py", line 1, in <module>
    from gmusicapi.clients.webclient import Webclient
  File "libs/gmusicapi/clients/webclient.py", line 6, in <module>
    from gmusicapi.clients.shared import _Base
  File "libs/gmusicapi/clients/shared.py", line 3, in <module>
    from gmusicapi.utils import utils
  File "libs/gmusicapi/utils/utils.py", line 44, in <module>
    log_filepath = os.path.join(my_appdirs.user_log_dir, 'gmusicapi.log')
  File "libs/appdirs.py", line 403, in user_log_dir
    version=self.version)
  File "libs/appdirs.py", line 356, in user_log_dir
    path = user_cache_dir(appname, appauthor, version)
  File "libs/appdirs.py", line 306, in user_cache_dir
    path = os.getenv('XDG_CACHE_HOME', os.path.expanduser('~/.cache'))
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/posixpath.py", line 268, in expanduser
    import pwd
  File "/Users/*/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/python/sandbox.py", line 898, in load_module
    raise ImportError('No module named %s' % fullname)
ImportError: No module named pwd
```

**Comment by simon-weber** [comment]
It doesn't look like there's an easy way around this because the appdirs call happens at the module-global level.

If you set up your own logging, you can probably get away with just removing that line, though.

**Comment by clocksmith** [comment]
Which line would I remove?

**Comment by simon-weber** [comment]
`log_filepath = os.path.join(my_appdirs.user_log_dir, 'gmusicapi.log')` in gmusicapi/utils/utils.py. You'll need to then set `debug_logging=False` when creating your Mobileclient ([docs](https://unofficial-google-music-api.readthedocs.org/en/develop/reference/mobileclient.html#gmusicapi.clients.Mobileclient.__init__)) and set up a logger manually if you want logging output. Here's some code to send the most important messages to stdout:

```python
    import logging
    logger = logging.getLogger('gmusicapi')

    important_handler = logging.StreamHandler()
    important_handler.setLevel(logging.WARNING)
    logger.addHandler(important_handler)

    formatter = logging.Formatter(
        '%(asctime)s - %(name)s (%(module)s:%(lineno)s) [%(levelname)s]: %(message)s'
    )
    important_handler.setFormatter(formatter)
```

**Comment by clocksmith** [comment]
Ok thanks. I am now getting another error which still falls under the umbrella of this issue:

Traceback (most recent call last):
  File "/Users/*/google-cloud-sdk/platform/google_appengine/google/appengine/runtime/wsgi.py", line 240, in Handle
    handler = _config_handle.add_wsgi_middleware(self._LoadHandler())
  File "/Users/*/google-cloud-sdk/platform/google_appengine/google/appengine/runtime/wsgi.py", line 299, in _LoadHandler
    handler, path, err = LoadObject(self._handler)
  File "/Users/*/google-cloud-sdk/platform/google_appengine/google/appengine/runtime/wsgi.py", line 85, in LoadObject
    obj = __import__(path[0])
  File "/Users/*/git/*/main.py", line 30, in <module>
    from gmusicapi import Mobileclient
  File "libs/gmusicapi/__init__.py", line 9, in <module>
    from gmusicapi.clients import Webclient, Musicmanager, Mobileclient
  File "libs/gmusicapi/clients/__init__.py", line 1, in <module>
    from gmusicapi.clients.webclient import Webclient
  File "libs/gmusicapi/clients/webclient.py", line 7, in <module>
    from gmusicapi.protocol import webclient
  File "libs/gmusicapi/protocol/webclient.py", line 18, in <module>
    from gmusicapi.protocol.shared import Call, authtypes
  File "libs/gmusicapi/protocol/shared.py", line 17, in <module>
    import requests
  File "libs/requests/__init__.py", line 58, in <module>
    from . import utils
  File "libs/requests/utils.py", line 26, in <module>
    from .compat import parse_http_list as _parse_list_header
  File "libs/requests/compat.py", line 42, in <module>
    from .packages.urllib3.packages.ordered_dict import OrderedDict
  File "libs/requests/packages/__init__.py", line 95, in load_module
    raise ImportError("No module named '%s'" % (name,))
ImportError: No module named 'requests.packages.urllib3'

**Comment by simon-weber** [comment]
huh, that's an odd one. requests [vendorizes urllib3](https://github.com/kennethreitz/requests/tree/master/requests/packages/urllib3), so I'm not sure how it could fail to find it. You could try debugging around the place where [requests tries to import it](https://github.com/kennethreitz/requests/blob/86f9b1616ae9fd0690a8a1746b9a8d111b4210d2/requests/packages/__init__.py#L64).

I also know urllib3 didn't support appengine until a recent version. Maybe try upgrading your version of requests?

**Comment by clocksmith** [comment]
Yes this is odd. I have `requests-2.6.0` and I can see `urllib3` in `requests/packages`. I'll see if I can find anything unusual around where requests imports it.

I wonder if app engine is just doing something weird not to include the `packages` under `requests`.
I have everything installed from pip, including `gmusicapi` under `/libs` and am using `sys.path.insert(0, 'libs')` to include things under `/libs`, but maybe it can't find things that are 2 levels deep?

I am entirely new to python and python app engine.

**Comment by simon-weber** [comment]
This issue hasn't been updated in a while, so I'm closing it. Feel free to reopen if you're still having trouble.

**Comment by xlorepdarkhelm** [comment]
In gmusicapi/clients/musicmanager.py
OAUTH_FILEPATH = os.path.join(my_appdirs.user_data_dir, 'oauth.cred')

Also needs to be changed to
OAUTH_FILEPATH = None  # os.path.join(my_appdirs.user_data_dir, 'oauth.cred')

Probably more needs to be done here to address this. I have oauth credentials, just not stored in a file, and it would be simpler for me to somehow not have to make a file for that.

**Comment by xlorepdarkhelm** [comment]
Next error:

Internal Server Error: /manager/testsongs
Traceback (most recent call last):
  File "/base/data/home/apps/s~gpm-playlist/1.388539967839736627/sitepackages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/base/data/home/apps/s~gpm-playlist/1.388539967839736627/manager/views.py", line 64, in testsongs
    api.login(user.email(), clear_passwd, gmusicapi.Mobileclient.FROM_MAC_ADDRESS)
  File "/base/data/home/apps/s~gpm-playlist/1.388539967839736627/sitepackages/gmusicapi/clients/mobileclient.py", line 57, in login
    raise OSError("a valid MAC could not be determined."
OSError: a valid MAC could not be determined. Provide an android_id (and be sure to provide the same one on future runs).


Cannot use gmusicapi.Mobileclient.FROM_MAC_ADDRESS

**Comment by xlorepdarkhelm** [comment]
Traceback (most recent call last):
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/sitepackages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/manager/views.py", line 64, in testsongs
    api.login(user.email(), clear_passwd, '364911a76fe0ffa1')  # gmusicapi.Mobileclient.FROM_MAC_ADDRESS)
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/sitepackages/gmusicapi/clients/mobileclient.py", line 64, in login
    if not self.session.login(email, password, android_id):
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/sitepackages/gmusicapi/session.py", line 170, in login
    res = gpsoauth.perform_master_login(email, password, android_id)
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/sitepackages/gpsoauth/__init__.py", line 66, in perform_master_login
    return _perform_auth_request(data)
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/sitepackages/gpsoauth/__init__.py", line 22, in _perform_auth_request
    headers={'User-Agent': useragent})
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/sitepackages/requests/api.py", line 109, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/sitepackages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/sitepackages/requests/sessions.py", line 468, in request
    resp = self.send(prep, **send_kwargs)
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/sitepackages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/base/data/home/apps/s~gpm-playlist/1.388540111994788285/sitepackages/requests/adapters.py", line 423, in send
    raise ConnectionError(e, request=request)
ConnectionError: HTTPSConnectionPool(host='android.clients.google.com', port=443): Max retries exceeded with url: /auth (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0xfcbbb890>: Failed to establish a new connection: [Errno 13] Permission denied',))


Maybe they are blocking this?

**Comment by simon-weber** [comment]
@xlorepdarkhelm 

> I have oauth credentials, just not stored in a file, and it would be simpler for me to somehow not have to make a file for that.

The best way to do this is to create and provide an `oauth2client.client.OAuth2Credentials` object. You can create them manually through google's client library. This is what the tests do (since the refresh token is stored encrypted in the travis settings, then hydrated from a string).

> Cannot use gmusicapi.Mobileclient.FROM_MAC_ADDRESS

That one is expected; a VPS-like environment can't provide a MAC address for you to use.

> ConnectionError: HTTPSConnectionPool(host='android.clients.google.com', port=443): Max retries exceeded with url: /auth (Caused by NewConnectionError(': Failed to establish a new connection: [Errno 13] Permission denied',))

This one is strange. Can you use `requests` to hit other https sites on gae?

**Comment by xlorepdarkhelm** [comment]
@simon-weber 

> The best way to do this is to create and provide an oauth2client.client.OAuth2Credentials object.
> You can create them manually through google's client library. This is what the tests do (since the
> refresh token is stored encrypted in the travis settings, then hydrated from a string).

Ok, I'll check that out. Thanks.

> That one is expected; a VPS-like environment can't provide a MAC address for you to use.

I figured as much, I found a different solution that looks like it should work.

> This one is strange. Can you use requests to hit other https sites on gae?

I will check it out. My concern is that Google is blocking access to that server from the GAE.

**Comment by xlorepdarkhelm** [comment]
It appears requests can't connect to SSL. Interesting.

**Comment by simon-weber** [comment]
It looks like requests/urllib has wrestled with app engine support in general. Maybe try requests 2.3.0 as per [this post](http://stackoverflow.com/a/28544305/1231454)?

**Comment by xlorepdarkhelm** [comment]
yeah, checking that now

**Comment by xlorepdarkhelm** [comment]
Ok, 2.3.0 seems to work for me.

**Comment by xlorepdarkhelm** [comment]
And I have success! I have been able to connect and query the Google Play Music API with Google App Engine.

**Comment by xlorepdarkhelm** [comment]
FYI -- your library rocks, for my purposes at least. Keep up the good work.

**Comment by simon-weber** [comment]
> success!

Nice! Glad everything is working.

> your library rocks

Thanks! :grin: Excited to see what you build.

I'm going to close this and open a new issue about updating the docs to contain GAE-specific instructions.

**Comment by xlorepdarkhelm** [comment]
Using your library with GAE to generate dynamically-updated playlists out of my music in Google Play, that pulls up particular selections of songs based on various conditions, to build an automatic daily playlist.

**Comment by simon-weber** [comment]
Oh, I'm actually working on the same problem, haha. I guess it goes to show it's a feature a lot of people want.

For what it's worth, it looks like we've got different approaches: mine's entirely clientside.

**Comment by xlorepdarkhelm** [comment]
Interesting, I want the client side to be there for customization of the process, but I want a server to automatically update the data without needing the client, which I am accomplishing through GAE's taskqueue system.

**Comment by xlorepdarkhelm** [comment]
Curiously enough, I have been able to load my entire library into a datastore Kind (NDB Model), and have it checking it every day for updates (minimizing my datastore writes as much as possible). I also have a way to retrieve these tracks to be available on the client. This all is important for me to be able to implement my additional tags I want for each track for managing the entire playlist. I still have my old code I've adapted multiple times for generating the playlist itself, and will roll that into place once I have a way to sort out my music.
